### DISCUSSION TITLE
Download link for the latest stable release

### DISCUSSION THREAD

**Discussion by solarkraft** [micropython/micropython#10519] [General]
Hi, I would like to build/modify a tool based on [ESP Web Tools](https://esphome.github.io/esp-web-tools/) to flash MicroPython using Chrome. I'm aware of the [BIPES tool](https://github.com/BIPES/BIPES/discussions/81), but this flashes an outdated version because it's specifically linked to one.

I would like to be able to always install the latest stable release so that this doesn't happen. For this I would need a stable way to obtain it. Is there an official way for this? Would you be willing to create one?

Having to scrape the page for the latest release would feel quite silly. Another shortcut, going by [the directory listing,](https://micropython.org/resources/firmware/) is forbidden. 

I believe that a Chrome-based flashing option would significantly lower the barrier to getting started with MicroPython - for flashing alone, but perhaps also for automatic provisioning (via [Improv Wi-Fi](https://www.improv-wifi.com/)/a custom solution). With WebSockets/[Weditor](https://github.com/micropython/micropython/discussions/9474) this could mean that if you have Chrome you could get started right away without installing anything. 

Compare that to the effort of installing esptool or one of the (frankly not very inviting looking or at all reliable) IDEs.

**Comment by jimmo** [comment]
@solarkraft This doesn't exactly answer your question but there was some discussion on this topic here: https://github.com/micropython/micropython/issues/8821

At the time I was hoping the OP would either agree with the suggestion we provided, or propose something else that suited their requirements.

**Reply by solarkraft** [reply]
Thanks for the link. 

I'm not a big fan of the suggestions. Parsing the page *might* work, but browsers tend to be restrictive about loading external resources, so it might not. At best it wouldn't be very elegant. 

A somewhat more elegant would be enabling the directory listing for `https://micropython.org/resources/firmware/`
 (and maybe doing something to make it downloadable from an external web page). This would allow for more flexibility, enabling me to allow the user to choose older versions. 

But in my book the "best bang for buck" solution would be a link like `https://micropython.org/resources/firmware/esp32c3-stable.bin` that points to the latest stable release. This would be universally and very easily accessible to any application that wants to fetch MicroPython firmware images (which I think there are quite a few of). Perhaps it could be generated as a symlink when the website is updated.

**Reply by scruss** [reply]
This hits a problem when a board had no stable release, as is the case for all boards added since 1.19.1 came out.

You might not be a fan of IDEs, but the author of Thonny has a method of checking for up-to-date releases. Maybe you could borrow that?

**Reply by jimmo** [reply]
> I'm not a big fan of the suggestions. Parsing the page might work

@solarkraft I'm not sure I follow. My suggestion was to provide a (machine readable) JSON or RSS feed. No parsing of pages required.

**Reply by solarkraft** [reply]
@jimmo You're right, I forgot about that. I think the main takeaway is that there aren't any really good options for us as users and something should be done to the website. 

I would be happy with the solution you suggest, the only caveat being that it's probably a little harder to implement.

**Reply by solarkraft** [reply]
> This hits a problem when a board had no stable release, as is the case for all boards added since 1.19.1 came out.

@scruss Maybe my terminology is wrong. What do you call the versions that are recommended as download on the website,  just "release"? That's what I mean.

**Reply by scruss** [reply]
"nightlies"? The releases correspond to a checkpoint of features and documentation. They're not always 100% stable on all platforms: the reason we have 1.19.1 as a release because 1.19 had some problems on some esp32 platforms.

See RP2040 [Pico](https://micropython.org/download/rp2-pico/) vs [Pico W](https://micropython.org/download/rp2-pico-w/). That latter only has nightly builds as it came out after the 1.19.1 release on June 18 2022.

Even more confusingly (but being addressed in the docs, thanks Jimmo!) is "latest", which is what you get if you pull directly from the repo. The latest documentation (the default if you go to [docs.micropython.org](https://docs.micropython.org/en/latest/)) doesn't correspond exactly to any available code, but gets synched at release time.
